text/0000-inferred-variant-constructor.md

```diff
@@ -0,0 +1,204 @@
+- Feature Name: `inferred_variant_constructor`
+- Start Date: 2017-03-11
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow the name (qualifier) of an enum variant to be elided in expressions and
+patterns whenever it can be inferred.
+
+# Motivation
+[motivation]: #motivation
+
+The main goal of this proposal is to improve ergonomics of enums in Rust by
+reducing verbosity.  Currently, there are two convenient ways to use enums.
+One would either:
+
+  - bring the enum type into scope (`use foolib::Foo;`), or
+  - bring the enum variants into scope (`use foolib::Foo::*;`).
+
+The former is less convenient than the latter, but the latter pollutes the
+namespace more.  One could always write enum variants fully qualified,
+`foolib::Foo::Alpha`, but that is even more tedious.
+
+This proposal intends to make variant expressions as convenient and
+pollution-free as field expressions today.  Whereas field expressions allow
+the user to extract fields of a struct without ever bringing the struct into
+scope, inferred variant constructors allow the user to construct/match
+variants of an enum without ever bringing the enum or its variants into scope.
+
+This could lower the usability barrier for enums, particularly for library
+users.  It would reduce the verbosity of enums to a level similar to
+dynamically typed languages, where strings (e.g. Python) or symbols
+(e.g. Lisps, Ruby) designate ad hoc variants
+
+The feature could encourage library writers to use domain-specific enums more
+often, rather than to re-purpose generic enums like `Option` or `Result`,
+leading to more readable code.  It would also allow library writers to use
```

Comment:
> so you won't need to use lots of booleans as flags to confuse the reader, but be able to make invocations like `LogBuilder::new().channels(5).with_verbosity(true).create()` instead of `Log::new(5,true))`.

Actually a strong reason for this RFC is to replace ad hoc enums such as that.  The ship has long sailed with regard to Rust’s context-sensitivity because of the `.field` syntax.  This is what enables “builder” patterns such as what you describe.

So in essence what I am proposing is to simply add first-class support for what folks are already doing anyway, but at least they can do it without all the boilerplate needed for the builder pattern.  They can also use a significantly more pleasant syntax:

~~~rust
Log {
    channels: 5,
    verbosity: _::Debug,
}
~~~